### ISSUE TITLE
🌟 [FEATURE] Training NequIP on Local Spectroscopic Properties (e.g., NMR parameters) for Specific Atoms Only

### ISSUE THREAD

**Issue by ozakary** [mir-group/nequip#487]
Hello,

I'm trying to train the NequIP architecture to predict local spectroscopic properties — specifically, the NMR isotropic magnetic shielding constant (***sigma_iso***) — for a subset of atoms (Xe atoms) within large structures (~1170 atoms per structure). ***sigma_iso*** is scalar and local (*i.e.*, depends only on the local atomic environment of the probe nucleus), so in principle, an energy per atom model should be well-suited for this task.

However, I would like the model to **only** learn from and predict ***sigma_iso*** values for Xe atoms by finding relations between Xe ***sigma_iso*** and its local atomic environment, while entirely ignoring all other atoms (H, C, N, O, F) during training, validation, and testing — since ***sigma_iso*** was only computed for Xe due to computational cost constraints.

---

## Dataset Format & Setup

My dataset is in `extxyz` format with this structure (snippet for illustration):

1170
Lattice="23.86466077 0.0 0.0 0.0 23.86466077 0.0 0.0 0.0 23.86466077" Properties=species:S:1:pos:R:3:tags:I:1:energies:R:1 original_dataset_index=0 energy=NaN free_energy=NaN pbc="T T T"
O 1.58842 0.345179 1.97149 3 NaN
C 1.65682 1.59036 1.25876 1 NaN
...
Xe 16.7214 14.5664 15.0416 5 5651.89752309
...
H 12.2018 15.8135 13.2785 0 NaN
Xe 17.757 16.9859 11.9365 5 5440.4428098
1170
Lattice="23.86466077 0.0 0.0 0.0 23.86466077 0.0 0.0 0.0 23.86466077" Properties=species:S:1:pos:R:3:tags:I:1:energies:R:1 original_dataset_index=1 energy=NaN free_energy=NaN pbc="T T T"
C 3.38732 3.08607 0.469146 1 NaN
C 2.54928 1.93986 2.83752 1 NaN
...
Xe 18.2989 16.9219 11.3747 5 5608.92690242
...
Xe 15.9836 13.2547 14.8303 5 5719.90265542
...
C 20.9991 23.5909 11.9565 1 NaN
H 20.8468 23.8836 20.4563 0 NaN
1170
Lattice="23.86466077 0.0 0.0 0.0 23.86466077 0.0 0.0 0.0 23.86466077" Properties=species:S:1:pos:R:3:tags:I:1:energies:R:1 original_dataset_index=2 energy=NaN free_energy=NaN pbc="T T T"
C 2.14485 0.195654 2.42156 1 NaN
O 3.26738 0.694005 1.75404 3 NaN
...
...

---
- Only Xe atoms have non-NaN per-atom "energy" (sigma_iso).
- The total energy of each structure is set to NaN since it should be irrelevant here.

---

### Model Builders:
```yaml
model_builders:
 - SimpleIrrepsConfig
 - EnergyModel
 - PerSpeciesRescale
 - RescaleEnergyEtc
```
### Dataset Definition:
```yaml
dataset: ase
dataset_file_name: ../dataset_7/2xe_cc3_tba_nmr_masked-atoms_sum.xyz
ase_args:
  format: extxyz
chemical_symbol_to_type:
  H: 0
  C: 1
  N: 2
  O: 3
  F: 4
  Xe: 5
```
### Loss & Metrics:
```yaml
loss_coeffs:
  total_energy:
    - 1.0
    - PerAtomMSELoss

metrics_components:
  - - total_energy
    - mae
    - PerAtom: True
  - - total_energy
    - rmse
    - PerAtom: True
```
---
## Problem

When running the training:

- All metrics (loss, MAE, RMSE) stay at zero throughout the entire training.
- It seems the model does not "see" or learn anything from the Xe per-atom values.
- I tried using the `mask_label` feature for `energies` (as suggested in [Issue #307](https://github.com/mir-group/nequip/issues/307)), but could not get it to work for this case — although the forces tutorial using [masks](https://github.com/mir-group/nequip/tree/masks/examples/mask_labels) worked.

---
## Question

Is there currently a supported way in NequIP (or Allegro) to:

- Train on per-atom properties (like sigma_iso)?
- Only for a subset of atoms (here: Xe atoms)?
- Ignore other atoms entirely from the loss function — treating their per-atom properties as NaN or irrelevant?

If not, would it be possible to extend the `mask_label` functionality to support per-atom energies in this context?

---
This feature would be highly valuable for the materials science and chemistry community, where local spectroscopies (NMR, EPR, Mössbauer, XAS, etc.) are often used.

---
Please let me know if you would need more details. Your guidance or suggestions on how to approach this problem using NequIP or Allegro would be highly appreciated!

**Comment by ozakary** [comment]
Hello,

This is a gentle follow-up regarding the enhancement issue posted earlier. As there has been no response so far, any feedback or guidance on the matter would be sincerely appreciated, especially given its relevance to ongoing research efforts.

Thank you for your time and consideration.

**Comment by cw-tan** [comment]
Hi @ozakary ,

Thanks for your interest. We've recently had the major [v0.7.0 update](https://github.com/mir-group/nequip/releases/tag/v0.7.0) and what you're describing is best implemented as a NequIP extension package, i.e. a separate repository that depends on NequIP, but manages a lot of its own code and custom functionality. A key example of a NequIP extension package is the [`allegro`](https://github.com/mir-group/allegro) package that implements the strictly local Allegro interatomic potential model within the NequIP framework (that covers the infrastructural aspects of training, testing, packaging, and compiling models). 

To proceed further on how you can implement all this in a NequIP extension package, please reach out to us on Zulip (register [here](https://nequip.zulipchat.com/)) or email [allegro-nequip@g.harvard.edu](mailto:allegro-nequip@g.harvard.edu). I believe that the functionality you want should be achievable in an extension package without modifications to the base `nequip` codebase.

**Comment by jvita** [comment]
@cw-tan Weighted per-atom loss functions would be very useful, as others have pointed out.

Prior to the v0.7.0 update I used to do this:
```python
class WeightedLoss:
    def __call__(
        self,
        pred: AtomicDataDict.Type,
        ref: AtomicDataDict.Type,
        key: str,
        mean: bool = True,
    ):
        """
        Args:
            pred: output of the model (can be a batch)
            ref:  training data (can be a batch)
            key:  which key to compute loss on, from the `loss_coeffs` config shown above.
            mean: whether to return a single scalar loss value.  Can basically be ignored (`assert mean`)
                as long as you don't want to use this loss as a metric (if you do, return instead a
                per-(atom/graph/whatever) tensor of losses).
        """
        assert mean
        weight = ref[WEIGHTS_KEY]
        pred = pred[key]
        ref = ref[key]
        term = ((pred - ref).abs()*weight[:, None]).mean()
        return term
```

However, with the update I see that the full `AtomicDataDict` is no longer passed into the `.update()` function, and instead only the specified `field` values are provided.

Before I go through the process of making a separate extension package, I just wanted to confirm that an approach like shown above is no longer possible.

**Comment by cw-tan** [comment]
Closing this issue now with mention that we now have extension package docs: https://nequip.readthedocs.io/en/latest/dev/extension_packages.html

and that the `MetricsManager` docstring should be a lot clearer about its behavior now https://nequip.readthedocs.io/en/latest/api/metrics.html#nequip.train.MetricsManager
